### PR TITLE
Convert function to TypeGuard

### PR DESCRIPTION
**Pull request by dwilt**
## Description

Right now, the function just returns a boolean. But we can leverage the generics provided to convert the function into [a TypeGuard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards).

See how it infers the type after the type guard is put in (`req.body` isn't `any` anymore):

## Checklist:

- [x] All tests are passing
- [x] Type definitions, if updated, pass both `test-ts-defs` and `test-ts-usage`
- [x] Benchmark performance has not significantly decreased
- [x] Bundle size has not been significantly impacted
- [x] The bundle size badge has been updated to reflect the new size

### CODE REVIEW

**Comment by dwilt** [comment]
@chrisbolin who should I request a review from?

**Comment by gksander** [comment]
@dwilt I like the idea of this a lot! I'm wondering tho, is there any way to avoid requiring a specific order for this to work? As you have it here, it requires your `any` value to be passed first to work as expected. Maybe there's a way for us to overload this function to handle both scenarios? I'll noodle on it a bit!

File: index.d.ts [hunk 1]
```diff
@@ -1,2 +1,2 @@
-declare function isEqual<A = any, B = any>(a: A, b: B): boolean;
+declare function isEqual<B = any>(a: any, b: B): a is B;
```
Comment (scottrippey) [inline index.d.ts:1]: Great idea, converting this to a typeguard!  Solid improvement.

It doesn't seem like there's any possible way to overload this function to return both `a is B & b is A`, since the typeguard syntax is very limiting.  

So the big question, as Grant said, is whether we should return `a is B` versus `b is A`.  This is purely a semantical choice, since the actual order of the params does not matter.

IMO, the current implementation, `a is B`, seems like the right choice.  If I was describing the logic in my own words, I'd say something like "if a is equal to b, then do stuff with a".  Basically, the first argument is the "subject", so it makes sense that I'd intend to interact with the subject next.

**Review by scottrippey — APPROVED** [review]

File: index.d.ts (hunk 1 above)
Comment (scottrippey) [reply index.d.ts:1]: @gksander  Should we consider this Type-only change a breaking change? Major, minor, or patch?

Comment (gksander) [reply index.d.ts:1]: That's a really good question 🤔 On the one hand, a change to TS types could theoretically break a `typecheck` process, but doesn't affect runtime behavior. My gut reaction is a minor release, but my opinions on that aren't terribly strong. What do you think?

Comment (scottrippey) [reply index.d.ts:1]: I think semver applies strongly to runtime behavior, but for static type checks, I prefer patch.  This can only change compile-time behavior, which is extremely easy to catch and fix, so it doesn't warrant any extra warnings.

Also, this typeguard change will result in **wider** types, because it implicitly widens the type of a variable, so typically that means no changes will be necessary.

The only thing I can think of, where this change could also require code changes, is for scenarios where you've already "widened" the scope manually (via a cast), and perhaps you have checks like [no-unnecessary-type-assertion](https://typescript-eslint.io/rules/no-unnecessary-type-assertion) which will then complain about the unnecessary cast.  This is actually ideal, since it will force you to remove the redundancy.

Long story short, this change should be really easy to adopt, so let's go with `patch` :)

Comment (gksander) [reply index.d.ts:1]: Love it. @dwilt will you rebase with (or merge in) `master` branch and then run `yarn changeset` and add a `patch` changeset to this PR? That way once we get this merged, we can release ASAP.

Comment (dwilt) [reply index.d.ts:1]: Done!

Comment (scottrippey) [inline index.d.ts:1]: I don't think there's any need for `= any` now, so I'm going to remove it.
```suggestion
declare function isEqual<B>(a: any, b: B): a is B;
```
